arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptySet.kt

```diff
@@ -0,0 +1,31 @@
+package arrow.core
+
+public class NonEmptySet<out T> private constructor(
+  private val elements: Set<T>
+) : Set<T> by elements {
+
+  public constructor(first: T, rest: Set<T>): this(setOf(first) + rest)
+
+  public operator fun plus(set: Set<@UnsafeVariance T>): NonEmptySet<T> = NonEmptySet(elements + set)
+
+  public operator fun plus(element: @UnsafeVariance T): NonEmptySet<T> = NonEmptySet(elements + element)
```

Comment:
There are other operations which you can implement more efficiently or with better types in `NonEmptySet`:
- `isEmpty() = false`, `isNonEmpty() = true`
- `any() = true` (the version which doesn't check a predicate)
- for `map`, you can implement a version which returns `NonEmptyList`

Reply:
`map` and `isEmpty` done! `isNotEmtpy()` and `any()` use `isEmpty` behind the scenes :slightly_smiling_face:

Reply:
Lovely! Thanks for putting all this love in the PR :)